client/src/main/java/io/pravega/client/stream/notifications/notifier/ScaleEventNotifier.java

```diff
@@ -0,0 +1,100 @@
+
+import io.pravega.client.state.StateSynchronizer;
+import io.pravega.client.stream.impl.ReaderGroupState;
+import io.pravega.client.stream.notifications.Listener;
+import io.pravega.client.stream.notifications.NotificationSystem;
+import io.pravega.client.stream.notifications.events.ScaleEvent;
+import javax.annotation.concurrent.GuardedBy;
+import lombok.Synchronized;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class ScaleEventNotifier extends AbstractEventNotifier<ScaleEvent> {
+
+    private static final int UPDATE_INTERVAL_SECONDS = Integer.parseInt(
+            System.getProperty("pravega.client.scaleEvent.poll.interval.seconds", String.valueOf(120)));
+    private final Supplier<StateSynchronizer<ReaderGroupState>> synchronizerSupplier;
+    private final AtomicBoolean pollingStarted = new AtomicBoolean();
+    @GuardedBy("$lock")
+    private ScheduledFuture<?> future;
+    @GuardedBy("$lock")
+    private StateSynchronizer<ReaderGroupState> synchronizer;
+
+    public ScaleEventNotifier(final NotificationSystem notifySystem,
+                              final Supplier<StateSynchronizer<ReaderGroupState>> synchronizerSupplier,
+                              final ScheduledExecutorService executor) {
+        super(notifySystem, executor);
+        this.synchronizerSupplier = synchronizerSupplier;
+    }
+
+    @Override
+    @Synchronized
+    public void registerListener(final Listener<ScaleEvent> listener) {
+        notifySystem.addListeners(getType(), listener, this.executor);
+        //periodically fetch the scale
+        if (!pollingStarted.getAndSet(true)) { //schedule the  only once
+            synchronizer = synchronizerSupplier.get();
+            future = executor.scheduleAtFixedRate(this::checkAndTriggerScaleNotification, 0,
+                    UPDATE_INTERVAL_SECONDS, TimeUnit.SECONDS);
+        }
+    }
+
+    @Override
+    @Synchronized
+    public void unregisterListener(final Listener<ScaleEvent> listener) {
+        notifySystem.removeListener(getType(), listener);
+        if (!notifySystem.isListenerPresent(getType())) {
+            cancelScheduledTask();
+            synchronizer.close();
+        }
+    }
+
+    @Override
+    @Synchronized
+    public void unregisterAllListeners() {
+        this.notifySystem.removeListeners(getType());
+        cancelScheduledTask();
+        synchronizer.close();
+    }
+
+    @Override
+    public String getType() {
+        return ScaleEvent.class.getSimpleName();
+    }
+
+    private void checkAndTriggerScaleNotification() {
+        this.synchronizer.fetchUpdates();
+        ReaderGroupState state = this.synchronizer.getState();
+        ScaleEvent event = ScaleEvent.builder().numOfSegments(state.getNumberOfSegments())
+                         .numOfReaders(state.getOnlineReaders().size())
+                         .build();
+        if (event.getNumOfReaders() > 0) { //Notify only if there are online readers in the reader group.
```

Comment:
I'm confused by this change. This is going to generate an event every time this method is called in the case there is at least one online reader, independent of a scaling event happening or not.

Reply:
To add to this, we need a explicit way of indicating that there has been a scaling event. The state of the reader group contains three sets: assigned, unassigned, future. A zk-like way of generating events would be to observe changes in these sets, but it might not be entirely trivial to do it in this case. The second way would be to have a list of explicit notifications in the state and have readers simply triggering locally based on changes to the notifications in the state.

Reply:
Had a chat with @tkaitchuck offline. We should issue an event every time the sum of assigned +  unassigned changes. In the absence of stream sealing, every segment either remains open, splits into multiple, or merges with some other segment. Consequently, the sum of assigned + unassigned cannot remain constant in the presence of scaling.

The observation about stream sealing is important because when a stream is sealed, we can hit the end of a segment and that segment has no successors, so we can have some anomalies where the number of segments remain constant even in the presence of scaling.

Another important observation is that this feature should actually be about the number of segments changing and not only about scaling. @tkaitchuck pointed out that the number can change also because, e.g., the config of a reader group is modified to add a new stream.

Please consider adding this feature to our documentation.

Reply:
Corrected the logic to generate a scale event. Updated the javadoc of `io.pravega.client.stream.notifications.ReaderGroupEventListener`.